src/hotkey.rs

```diff
@@ -0,0 +1,269 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Hotkeys and helpers for parsing keyboard shortcuts.
+
+use crate::{KeyCode, KeyEvent, KeyModifiers};
+
+/// A description of a keyboard shortcut.
+///
+/// This type implements `PartialEq<KeyEvent>`, and it is only intended to
+/// be used to describe shortcuts, and recognize them when they arrive.
+///
+/// # Examples
+///
+/// [`SysMods`] matches the Command key on macOS and Ctrl elsewhere:
+///
+/// ```
+/// use druid::{HotKey, KeyEvent, RawMods, SysMods, KeyCode};
+///
+/// let hotkey = HotKey::new(SysMods::Cmd, "a");
+///
+/// #[cfg(target_os = "macos")]
+/// assert!(hotkey == KeyEvent::for_test(RawMods::Meta, "a", KeyCode::KeyA));
+///
+/// #[cfg(target_os = "windows")]
+/// assert!(hotkey == KeyEvent::for_test(RawMods::Ctrl, "a", KeyCode::KeyA));
+/// ```
+///
+/// `None` matches only the key withuot modifiers:
+///
+/// ```
+/// use druid::{HotKey, KeyEvent, RawMods, SysMods, KeyCode};
+///
+/// let hotkey = HotKey::new(None, KeyCode::ArrowLeft);
+///
+/// assert!(hotkey == KeyEvent::for_test(RawMods::None, "", KeyCode::ArrowLeft));
+/// assert!(hotkey != KeyEvent::for_test(RawMods::Ctrl, "", KeyCode::ArrowLeft));
+/// ```
+///
+/// [`SysMods`]: enum.SysMods.html
+#[derive(Debug, Clone)]
+pub struct HotKey {
+    mods: RawMods,
+    key: KeyCompare,
+}
+
+/// Something that can be compared with a keyboard key.
+#[derive(Debug, Clone, PartialEq)]
+pub enum KeyCompare {
+    Code(KeyCode),
+    Text(&'static str),
+}
+
+impl HotKey {
+    /// Create a new hotkey.
+    ///
+    /// The first argument describes the keyboard modifiers. This can be `None`,
+    /// or an instance of either [`SysMods`], or [`RawMods`]. [`SysMods`] unify the
+    /// 'Command' key on macOS with the 'Ctrl' key on other platforms.
+    ///
+    /// The second argument describes the non-modifier key. This can be either
+    /// a `&'static str` or a [`KeyCode`]. If it is a `&str`, it will be compared
+    /// against the [`unmodified text`] for a given key event; if it is a [`KeyCode`]
+    /// it will be compared against the event's key code.
+    ///
+    /// In general, [`KeyCode`] should be preferred for non-printing keys, like
+    /// the arrows or backspace.
```

Comment:
This docstring leaves me not certain whether Shift-A is `(RawMods::Shift, "a")` or `(RawMods::Shift, "a")`. I see a TODO below; that seems reasonable and we can get back to this (especially win vs mac).

Reply:
yea I'm personally confused about this situation right now. On mac, shift is applied, so 'unmod text' is "A" for shift+a, and "a" for ctrl+a. We could case normalize, but I haven't thought through all the edge cases.